### ISSUE TITLE
nscde_fvwmclnt fails to start programs from infostore

### ISSUE THREAD

**Issue by cvjb** [NsCDE/NsCDE#88]
With NsCDE 2.0 starting programs that are defined via infostore, e.g. the calculator, fails. Hardocded programs like e.g. vlc are fine.
In the FVWM Diagonotics Console (aka "Watch Errors") the following is logged when trying to start e.g. the calculator:

`[fvwm][__execute_function]: <<ERROR>> No such command 'xcalc'`

I could fix this by replacing "$@" by "$*" in the nscde_fvwmclnt script (line 51):

original version:  `exec $FVWM_CLNT "$@"`
new version:  `exec $FVWM_CLNT "$*"`

System information:
OS: Opensuse Tumblewheed (20211228)
NsCDE: Git clone, sha1: 863f2a1ea62528c8e2cf2d6375027575b5703c23
ksh: sh (AT&T Research) 93u+ 2012-08-01; Package from official OS-repo

**Comment by NsCDE** [comment]
Hi @cvjb

This is very similar to the second current issue we have here. I just booted Opensuse Tumblewheed VM and tried to reproduce this. I can't.

You didn't actually wrote HOW do you run this. nscde_fvwmclnt from the terminal or what? How can I reproduce this? I cannot get no such command error from fvwm __execute_function.

I see you found a fix for yourself, but this is far from finished. Please let me know how to trigger this. How do you start the programs to fail like that? Menu, Submenu, Front Panel? nscde_fvwmclnt from a terminal  window? I cannot get it.

And please give me:

grep ExecUseShell /usr/share/NsCDE/fvwm/Main.fvwmconf

**Comment by cvjb** [comment]
Hi @NsCDE 

> You didn't actually wrote HOW do you run this. nscde_fvwmclnt from the terminal or what? How can I reproduce this? I cannot get no such command error from fvwm __execute_function.
> I see you found a fix for yourself, but this is far from finished. Please let me know how to trigger this. How do you start the programs to fail like that? Menu, Submenu, Front Panel? nscde_fvwmclnt from a terminal window? I cannot get it.

I have started the programs from the subpanels. The calculator e.g. from the 'Tools' subpanel. So nsdc_fvwmclnt is called through the generated entries of the subpanels.

The output 'No such command ...' was then shown in the FvwmConsole opened through 'Watch Errors' (started from the 'Desktop Settings' subpanel), then opening the 'Fvwm Diagnostics Console' via the 'Window menu' or simply by executing 'FvwmCommand FvwmConsole' from an xterm window.

But! This is not always the case. I do not know why, sometimes that error is printed to the console and somtimes it is not. I have no idea what is the reason for this. For this reply I have reverted my change in nscde_fvwmclnt, but I only get the 'No such command'-message for e.g. the 'Volume Control' entry in the 'Graphics and Multimedia' subpanel, the 'Webbrowser' entry in the 'Communications' subpanel, and the 'NsCDE Process Manager' in the 'Desktop Settings' subpanel. For the calculator entry in the 'Tools' subpanel nothing is logged in the console.

> And please give me:
> 
> grep ExecUseShell /usr/share/NsCDE/fvwm/Main.fvwmconf

Sorry I have missed the maybe important point, that NsCDE is installed below '/opt/NsCDE' on my system. Here is the what grep has found:

```
~ $ grep ExecUseShell /opt/NsCDE/share/NsCDE/fvwm/Main.fvwmconf
ExecUseShell /usr/bin/ksh93
```

This is how I installed NsCDE:

1. git clone ...
2. cd ./NsCDE
3. ./configure --prefix=/opt/NsCDE
4. make
5. sudo make install

Do you need the config.log?

At the first start-up there was no ~/.NsCDE directory.

I am starting NsCDE through an ~/.xinitrc file using XDM as login manager. In the .xinitrc NsCDE is started by calling

```
dbus-launch --exit-with-session /opt/NsCDE/bin/nscde
wmpid=$!
wait $wmpid
```

Do you need anything else, please feel free to ask. What I could not get is a debug/trace output of nscd_fvwmclnt when called via the subpanels. I had added 'set -xv' to the script, but the output did not show anywhere I have looked for, e.g. ~/.xsession-errors-:0 or the system journal.

**Comment by NsCDE** [comment]
Hi @cvjb 

Tell me please: if using non-standard non-system paths like you described, do you have /opt/NsCDE/bin in your $PATH prior to starting X? Maybe FVWM 'Exec' function cannot find nscde_fvwmclnt.

You said that ~/.NsCDE didn't existed prior to start. Did you passed initial setup wizard with the questions? Did you selected then some theme and restarted NsCDE? Was that all (visually) passed well?

Somehow, I cannot belive that "$@" vs "$*" in script difference matters. I have tested this on many Linux and BSD systems, and never got it non-working. Did you edited Tools menu in ~/.NsCDE manually? Are the calls to nscde_fvwmclnt quoted? As well as FvwmCommand (for which versions 2 and 3 nscde_fvwmclnt is a wrapper / proper distributor), nscde_fvwmclnt expects commands to be quoted properly. Best thing is to use single quotes.

Please send here your ~/.NsCDE/Subpanels.actions.

And last ...
Can you try regular RPM for your system instead (make uninstall before that to avoid mess)
https://github.com/NsCDE/NsCDE/releases/download/2.0/NsCDE-2.0.3.x86_64.suse.rpm

**Comment by cvjb** [comment]
/opt/NsCDE/bin is in the PATH variable that fvwm uses. I am setting it explicitly in the .xinitrc.

The initial setup wizard started and I did pass (answer) all questions. I also restarted NsCDE at the end. All looks good. The front panel is there with all the expected buttons and subpanels. There were and are no issues shown or seen.

"$@" and "$*" makes a difference if there are more than one parameter given to a shell script, which is the case in the calls with the infostore variables. I have now tried the following calls from the shell prompt to resemble the actual call made in nscde_fvwmclnt:

1. `FvwmCommand "Exec" "exec" "$[infostore.calculator]"` ==> Nothing happens, no calculator. In an open FvwmConsole the following error is logged: `[fvwm][__execute_function]: <<ERROR>> No such command 'xcalc'`.
2. `FvwmCommand Exec exec $[infostore.calculator]` ==> Nothing happens, no calculator. In an open FvwmConsole the following error is logged: `[fvwm][__execute_function]: <<ERROR>> No such command 'xcalc'`. Just my normal paranoia ;-) I did not expect a difference to the first version.
3. `FvwmCommand "Exec exec $[infostore.calculator]"` ==> Calculator starts. 

Background: `"$@"` will create one argument for each parameter given to the shell script. That is why I enclosed each string with double quotes in the first call. `"$*"` will create one single argument holding **all** parameters given to the shell script.

FvwmCommand takes each parameter as a single fvwm command. Quotes have to be used to send commands including spaces, e.g. the above 'Exec exec ...'. is one command. So I think nscde_clnt should use `"$*"` instead of `"$@"`.

But I will try the official rpm and report the results.

**Comment by cvjb** [comment]
> But I will try the official rpm and report the results.

No changes in the behaviour with the rpm installation on my system. All entries of the subpanels using the nscde_fvwmclnt script fail. In an open FvwmConsole an error is logged: `[fvwm][__execute_function]: <<ERROR>> No such command '...'`.

I have removed my installation in `/opt/NsCDE` and the local configuration below `~/.NsCDE`. The installation was executed from the text console. So NsCDE was not running while it was installed anew.

After the installation a `type nscde' in bash or `whence nscde` in ksh93 returns `/usr/bin/nscde`. I am also explicitly calling `/usr/bin/nscde` in my `.xinitrc`.

Changing the `"$@"` in nscde_fvwmclnt to `"$*"` fixes the issue.

Please, feel free to ask me to check or test anything else if needed.

**Comment by NsCDE** [comment]
Hi @cvjb 

While I agree that $* is more robust, and may be introduced as the result of this issue, there is a big mistery why on your system it behaves in that way, while on mine everything is ok without changing $@ into $* in nscde_fvwmclnt.

I must solve this, because similar cases, or future changes should probably take into consideration what we find here. It is impossible that my VM with this exact system is behaving ok on this same actions, while yours behaves as if there is some other, icompatible calling shell in place. There has to be some explanation for this.

Please give this info exactly as it is:

- rpm -q ksh bash
- ls -l /bin/sh
- ls -l /etc/alternatives/sh

- If /bin/sh actually points to bash, have you changed anything in /etc/bash.bashrc? /etc/ksh.kshrc is a symlink to bash.bashrc as I see ...

- Do you have any ~/.profile ~/.kshrc in $HOME? Also ~/.bashrc and ~/.bash_profile. Can you try to move them away temporary and re-login? Maybe some shopt, IFS or some other variable or similar is changing default behaviour of the Korn shell.

- Try to put "-p" at the end of "#!/usr/bin/ksh93" shebang in nscde_fvwmclnt and see if there is a change. Of course, you have to revert temporary to $@ at the end of the nscde_fvwmclnt.

Then please call programs which are referenced by infostore from subpanels and panel, and please try also this from the command line and paste here is there is any output on the terminal, or in the debug log of fvwm.

nscde_fvwmclnt 'Exec exec $[infostore.terminal]'
nscde_fvwmclnt 'Exec exec $[infostore.calculator]'
nscde_fvwmclnt 'Exec exec xterm'

**Comment by cvjb** [comment]
Hi @NsCDE

Yes the different behaviour between the two installations is very strange.

Here the data you have asked for:

```
$ rpm -q ksh bash
ksh-93vu-5.4.x86_64
bash-5.1.12-4.1.x86_64
```

```
$ ls -l /bin/sh
lrwxrwxrwx 1 root root 4  7. Dez. 00:05 /bin/sh -> bash
```

```
$ ls -l /etc/alternatives/sh
lrwxrwxrwx 1 root root 13 24. Aug 2020  /etc/alternatives/sh -> /usr/bin/bash
```

I have not changed anything in `/etc/bash.bashrc` and there is also no `/etc/bash.bashrc.local`.

In my home directory there are:
* `.profile`
* `.bash_profile`
* `.bashrc`

`IFS` is not changed in one of these files.
The following shopt calls are defined in the `.bashrc`:
```
shopt -s checkwinsize
shopt -s autocd
```

There is no `.kshrc`.

Where can I found the debug output of fvwm? Is this the logging seen in the FvwmConsole?

Before temporarily removing the profile/ressource-files for bash from the home directory I have added the "-p" to the shebang of the `nscde_fvwmclnt` script and also reverted "$*" back to "$@". So the script now looks as the following:
```
$ head -5 /usr/bin/nscde_fvwmclnt
#!/usr/bin/ksh93 -p

#
# This file is a part of the NsCDE - Not so Common Desktop Environment
# Author: Hegel3DReloaded
$ tail -5 /usr/bin/nscde_fvwmclnt
   exec $FVWM_CLNT "$@"
else
   echo "ERROR: Cannot find suitable FvwmCommand for your setup." >&2
fi
```

No changes in the behaviour after that. The FvwmConsole still logs the 'No such command ...' message, when trying to start e.g. the calculator from the 'Tools'-subpanel. Programs started from the panel or the menu, e.g. "Applications -> Dienstprogramme -> Calculator" are fine. Sorry, for the german "Dienstprogramme", my locale is set to de_DE.UTF-8.

All three calls of nscde_fvwmclnt from the command line are working as expected. The programs are started. But this does also work without the "-p" after the shebang in the script.

My default shell is bash. As expected it makes a difference using single or double quotes in these calls:
```
$ /usr/bin/nscde_fvwmclnt "Exec exec $[inforstore.calculator]"
-bash: inforstore.calculator: syntax error: invalid arithmetic operator error token is "calculator")
```
and
```
$ /usr/bin/nscde_fvwmclnt \"Exec exec \$\[infostore.calculator\]\"
```
will not open the calculator, but again logs the 'No such command ...' in the FvwmConsole.
```
$ /usr/bin/nscde_fvwmclnt "Exec exec \$[infostore.calculator]"
```
will then open xcalc.

I have then changed the shell in the same terminal to ksh93 and tried the following:
```
$ /usr/bin/nscde_fvwmclnt "Exec exec $[inforstore.calculator]"
```
That is opening xcalc as expected.

So ksh93 and bash show differences in the handling of the double quotes in these calls. I have tried the calls as douvle quotes are used in e.g. `/usr/share/NsCDE/defaults/Subpanel.actions` or the generated subpanel-action files in `~/.NsCDE/`.

I will now remove the local profiles and resource files in my home directory and will test the calls again. To shorten this already long answer, I will post the results in a new reply.

**Comment by cvjb** [comment]
Well, the result will be shorter than I thougt.

No changes in the behaviour even without any personal profile and resource files.

After that test I have then changed my login shell from `/bin/bash` to `/usr/bin/ksh93`. No changes either.

What about changing `sh` to `/usr/bin/ksh93` via `/etc/alternatives/sh`?

**Comment by NsCDE** [comment]
Hi @cvjb

Your results from the command line are expected and same as mine. The real question is why you get "No such command" from the click to the item in the subpanel, and not by doing this from the menu. Even with default clean subpanels configuration.

BTW, no need to change login shell to ksh, or to relink alternatives. A whole bunch of other systems are using sh or bash as /bin/sh and this doesn't bother NsCDE.

Please give me:
`grep calculator /usr/share/NsCDE/defaults/Subpanels.actions`

If nothing is changing that file, output should look exactly like this:
```
S4,ENTRY,$[gt.Calculator],,NsCDE/32x32/apps/accessories-calculator.png,nscde_fvwmclnt \"Exec exec $[infostore.calculator]\"
```
and this:
`grep calculator ~/.NsCDE/Subpanels.fvwm2.fvwmgen`

it should be:

```
Test (x $[infostore.calculator]) *NsCDE-Subpanel4: \
 Icon "NsCDE/32x32/apps/accessories-calculator.png", \
 Action (Mouse 1) f_SubpanelItemExec NsCDE-Subpanel4 Btn-4-13 Exec exec $[infostore.calculator], \
```

Next, write "Exec exec $[infostore.calculator]" in ~/some_test.conf (without quotes) and read it with FVWM Read command:
```
nscde_fvwmclnt 'Read $[HOME]/some_test.conf'
```
... calculator should appear, there is no any "No such command" when I do this.

P. S.
You asked for debugging output: if you don't see Xorg stderr and stdout where fvwm also logs (watch errors), I must say that lightdm used on OpenSuse system puts this in /dev/null unfortunately no matter if it says "Logging to .xsession-errors" in the /var/log/lightdm/lightdm.log, and I don't see the way how to fix it (bug?). However, you can cheat it: go to the "Exit" on the Front Panel, and from System Action Dialog choose "X Window System: Failsafe (xterm)". You will be left without window manager and one xterm will appear in the upper left corner of the screen. Do not close it, it holds your whole X session. In that xterm, you can type "nscde". New NsCDE session will start, FVWM will again manage your screen and all what is logged by NsCDE and FVWM will be visible in this xterm.

**Comment by NsCDE** [comment]
@cvjb 

I was finally able to reproduce behaviour you are describing here. Old (valid) Subpanel entries were preventing me to see the bug. This is not only on your system, it is the problem with quoting when subpanel entries are generated. Replacing "$@" with "$*"  workarounds this, but I will like to do it in a "hard" way - fix the actual cause of the problem (it was introduced recentry) and keep argv parsing of the nscde_fvwmclnt as much as close to FvwmCommand.

I will be back when I fix this in a proper way.

**Comment by NsCDE** [comment]
I have commited the fix. Subpanel config generator will not lose backslashes before double quotes anymore. You can pull the master, or simply update (rpm -Fvh) freshly built RPM (2.0-5) which I have attached in the assets of 2.0.

After this, open ANY subpanel, click on the leftmost menu button and choose "Reload Subpanel" - that will regenerate your ~/.NsCDE/Subpanels.fvwmX.fvwmgen and reload subpanels with the new configuration.

**Comment by cvjb** [comment]
Great! It works. I have tested two scenarios:

1. Updated via the 2.0-5 rpm and reload of the panels. The calculator and e.g. pavucontrol are starting via the subpanels.
2. I have removed the rpm installation of NsCDE and pulled the newest code version from the repo. Then I have again configured the installation to be below `/opt/NsCDE` and called `make` and `sudo make install`. That variant does also work as expected.

Thank you!
